source/agora/consensus/EnrollmentManager.d

```diff
@@ -168,8 +168,11 @@ public class EnrollmentManager
 
         // load the count of 'populate' of `PreImageCycle`
         const uint populate_count = this.getCycleIndex();
+
+        // if there exists an erollment key, `populate_count` must be above 0.
+        assert(this.enroll_key == Hash.init || populate_count > 0);
         foreach (_; 0 .. populate_count)
-            this.cycle.populate(this.key_pair.v, true);
+            this.cycle.populate(this.key_pair.v, this.enroll_key, true);
```

Comment:
I see now why you are loading it in the  constructor. I think we should rather move this outside of the constructor, so pre-images only get generated when needed (and possible). Otherwise the first time a validator runs, tons of pre-images will be generated for nothing.